src/box/lua/connpool.lua

```diff
@@ -271,46 +329,111 @@ local function filter(opts)
             table.insert(static_candidates, instance_name)
         end
     end
+
     -- Return if retrieving dynamic information is not required.
-    if next(static_candidates) == nil or next(dynamic_opts) == nil then
+    if mode == 'any' then
         return static_candidates
     end
 
     -- Filter the remaining candidates after connecting to them.
-    local connected_candidates = connect_to_candidates(static_candidates)
+    local connected_instances = acquire_all_candidates(static_candidates)
+
     local dynamic_candidates = {}
-    for _, instance_name in pairs(connected_candidates) do
+    for _, instance_name in pairs(connected_instances) do
         if is_candidate_match_dynamic(instance_name, dynamic_opts) then
             table.insert(dynamic_candidates, instance_name)
         end
     end
+
     return dynamic_candidates
 end
 
+-- Subroutine for the `call()` function. Tries to find an active
+-- connection matching the requirements and considering the
+-- specified priorities.
 local function get_connection(opts)
-    local mode = nil
+    assert(opts.mode == 'alive' or opts.mode == 'ro' or opts.mode == 'rw' or
+           opts.mode == 'prefer_ro' or opts.mode == 'prefer_rw')
```

Comment:
Why don't we separate the `is_alive` option from the mode? Then we could have something like `opts = {mode = 'rw', is_alive = true}` or `opts = {mode = 'prefer_ro', is_alive = true}`. For me this makes it a lot more flexible for the user and a lot more easy to understand.

Reply:
`mode` = `rw` has no sense if `is_alive` = `false`. The same for `ro`, `prefer_*` -- all the mode values. So, I'm looking on it this way:

* `mode` = `any` -- no constraints
* `mode` = `alive` -- at least alive
* `mode` = `ro`/`rw` -- alive and has the given mode
* `mode` - `prefer_ro`/`prefer_rw` -- the same as previous, but with the give priority

Reply:
We discussed it with @grafin and @georgiy-belyanin today. A summary is below.

* `connpool.call()` doesn't require any changes. We agreed to don't introduce any new options or option's values to it.
* The `connpool.filter()` behavior may be changed to return all alive instances by default. The experimenal status of the module allows it and we don't expect problems from this change. Moreover, we guess that this behavior is likely what early adopters actually expect.
* A separate option to return all instances disregarding its connection status is to be introduced for `connpool.filter()`.
* We considered several options:
  * Add it to the `mode` enumeration.
  * Add it into a newly created `ready` enumeration (to add more interesting readiness checks later to it).
  * Add a separate boolean option.
* The latter option looks the most safe and straightforward. We agreed on it.
* The name may be `require_connected` or `require_alive`[^1]. Not necessary some ideal name, because later we can revisit it in context of a more general readiness checking solution -- and deprecate this option.

[^1]: After the discussion I thought that it is more intuitive, when the default is `false`, so maybe it worth to consider `allow_disconnected` or `skip_connection_check`.

Reply:
Thank you for writing down the points and options we've discussed. I've implemented the suggested behaviour using the `skip_connection_check` option.

Reply:
Seems resolved, so I'm going to close this discussion.